RiotSwiftUI/Modules/Template/SimpleUserProfileExample/ViewModel/TemplateUserProfileViewModel.swift

```diff
@@ -0,0 +1,87 @@
+// 
+// Copyright 2021 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import SwiftUI
+import Combine
+    
+@available(iOS 14.0, *)
+class TemplateUserProfileViewModel: ObservableObject, TemplateUserProfileViewModelProtocol {
+    
+    // MARK: - Properties
+    
+    // MARK: Private
+    private let userService: TemplateUserProfileServiceProtocol
+    private var cancellables = Set<AnyCancellable>()
+    
+    // MARK: Public
+    @Published private(set) var viewState: TemplateUserProfileViewState
+    
+    var completion: ((TemplateUserProfileViewModelResult) -> Void)?
+    
+    private static func defaultState(userService: TemplateUserProfileServiceProtocol) -> TemplateUserProfileViewState {
+        return TemplateUserProfileViewState(avatar: userService.avatarData, displayName: userService.displayName)
+    }
+    
+    // MARK: - Setup
+    init(userService: TemplateUserProfileServiceProtocol, initialState: TemplateUserProfileViewState? = nil) {
+        self.userService = userService
+        self.viewState = initialState ?? Self.defaultState(userService: userService)
+        
+        userService.presencePublisher
+            .map(TemplateUserProfileStateAction.updatePresence)
+            .receive(on: DispatchQueue.main)
```

Comment:
The current design is that all dispatches happen on the main thread. This simplifies the implementation and will usually meet the requirements for most ViewModels. As we get into implementing larger/more complex views we will likely need to do this work off them main thread(a list like room bubbles where we may do iteration/filtering etc across collections).

We could move towards an implementation like mvrx that the android team use. Their 'withState'/setState approach does the the work in the background and in a thread safe way. This approach/api may also suit async/await and we could find good commonality across the platforms with an mvrx type approach to view models and using kotlin coroutins/ swift async/await.

Reply:
Again similar to the comment below, if we have the common pieces in a superclass we can make this syntactically nicer and be able to do something like:

```swift
userService.presencePublisher
  .map(TemplateUserProfileStateAction.updatePresence)
  .dispatchActionTo(self)
```